src/platform/nrfconnect/wifi/WiFiManager.cpp

```diff
@@ -504,12 +512,36 @@ void WiFiManager::ConnectHandler(Platform::UniquePtr<uint8_t> data)
 
 void WiFiManager::DisconnectHandler(Platform::UniquePtr<uint8_t>)
 {
+    if (Instance().mDisconnectRequested)
+    {
+        Instance().mDisconnectRequested = false;
+
+        if (Instance().mRecoveryStarted)
+        {
+            Instance().AbortConnectionRecovery();
+        }
+    }
+    else
+    {
+        // Workaround: schedule the application level connection recovery in kSupplicantReconnectionTimeoutMs to give WPA supplicant
+        // some time to restore it.
+        if (!Instance().mRecoveryStarted)
+        {
+            Instance().mRecoveryStarted = true;
+            DeviceLayer::SystemLayer().StartTimer(
+                System::Clock::Milliseconds32(kSupplicantReconnectionTimeoutMs),
```

Comment:
I think we can save some time if we cancel this timer in the `WiFiManager::ConnectHandler` right away (now we need to wait for another `Recover()` call to issue `AbortConnectionRecovery()`).

Reply:
No, it will be aborted in the `ScanDoneHandler` after checking if SSID was found: https://github.com/nrfconnect/sdk-connectedhomeip/pull/275/files#diff-360bd1fbca9712db3463cc36c37be05596a04f08eab2f947cee8bf1c07ad6835R408

Reply:
Yes but... the original recovery mechanism was introduced to have a chance to find the SSID which is not be present immediately after reboot for whatever reason (i.e. it does not care about the network credentials). Here, we are handling the connection instabilities at *runtime*, so I would expect the recovery to be finalized only when the connection is back again instead of relying solely on the SSID being found.

Reply:
Discussed F2F, we will create a follow-up task for that, which could be connected with the WiFiManager refactor.